### PR TITLE
Assistant: load additional file with kinematics parameters

### PR DESCRIPTION
**Pull request by simonschmeisser**
### Description

This adds an option in MoveIt setup assistant to setup an additional yaml file with parameters for the kinematic solver(s). The main motivation for this is to improve integration of `opw_kinematics` [1] via the `moveit_opw_kinematics_plugin`[2]. Currently the kinematic chain needs to be described in parameters in the `kinematics.yaml` file. This makes it inconvenient to reuse the information from outside MoveIt (e.g. `descartes`) as other setups typically have no access to the group names defined in SRDF. This and alternative approaches was discussed at length in [3].

The plan is to distribute a MoveIt - independent YAML file along with the URDF description as can be seen in [4]. The user would then select the YAML file distributed along with the URDF file by ROS-Industrial. A separate entry is added in launch/planning_scene.launch per group to load this parameter file.

Adding this feature would allow dropping ikfast generated code for ros-i repositories, significantly reducing maintenance effort by going from 1k+ lines of code per robot to 9 lines of parameters.

grafik

1: https://github.com/Jmeyer1292/opw_kinematics
2: https://github.com/JeroenDM/moveit_opw_kinematics_plugin
3: https://github.com/ros-industrial/fanuc/issues/245
4: https://github.com/ros-industrial/fanuc/pull/284

@gavanderhoorn @JeroenDM 

As I'm currently running melodic-devel and would hope for this to be backported, this PR is against melodic-devel. Once discussed I can rebase against master if this is more convenient for whoever does the merge.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Thanks for submitting this PR @simonschmeisser.

Just to make it extra clear: the functionality contributed is not OPW specific, correct? It is essentially a way to configure an additional `.yaml` file to be loaded, next to MoveIts default `kinematics.yaml`.

This mechanism could be used by / with other plugins as well, it seems?

**Comment by simonschmeisser** [comment]
> This mechanism could be used by / with other plugins as well, it seems?

Yes of course, thanks for pointing this out. This is completely agnostic of course and could be used with other plugins that require configuration (eg `caching_ik_plugin`) or also for the likes of UR (and "clones" like Techman) that have an configurable analytical solver

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 1]
```diff
@@ -1140,6 +1153,9 @@ bool ConfigurationFilesWidget::addTemplateString(const std::string& key, const s
 // ******************************************************************************************
 bool ConfigurationFilesWidget::copyTemplate(const std::string& template_path, const std::string& output_path)
 {
+  // clear template strings in order to update them
+  template_strings_.clear();
+
   // Check if template strings have been loaded yet
   if (template_strings_.empty())
```
Comment (v4hn) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: It does not make sense to clear the list and then condition on whether it is empty.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch [hunk 2]
```diff
@@ -11,6 +11,7 @@
   <node name="$(anon rviz)" launch-prefix="$(arg launch_prefix)" pkg="rviz" type="rviz" respawn="false"
 	args="$(arg command_args)" output="screen">
     <rosparam command="load" file="$(find [GENERATED_PACKAGE_NAME])/config/kinematics.yaml"/>
+    [KINEMATICS_PARAMETERS_FILE_NAMES_BLOCK]
```
Comment (v4hn) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch]: Actually, I believe the kinematics configuration should be properly loaded from the `/move_group` namespace, also in RViz. I would prefer to see this line (and the one above) removed from here unless you can confirm it is still needed.

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 1 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: True, I'll remove the conditional (and caching) if that's ok?

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch [hunk 3]
```diff
@@ -16,6 +16,7 @@
   <!-- Start Benchmark Executable -->
   <node name="$(anon moveit_benchmark)" pkg="moveit_ros_benchmarks" type="moveit_run_benchmark" args="$(arg cfg) --benchmark-planners" respawn="false" output="screen">
     <rosparam command="load" file="$(find [GENERATED_PACKAGE_NAME])/config/kinematics.yaml"/>
+    [KINEMATICS_PARAMETERS_FILE_NAMES_BLOCK]
```
Comment (v4hn) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch]: See above?

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 1 above)
Comment (v4hn) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: Why did you disable the caching in the first place?
I don't know the speedup is really significant, but I don't see why it should break your new feature?

**Comment by JeroenDM** [comment]
Funny coincidence, I actually started implementing something to address the same problem this weekend and planned to post something about it today. :)

But you're approach is way more generic. I experimented with adding extra attributes to [GroupMetaData](https://github.com/ros-planning/moveit/blob/379acf91aa3630f320db77ba85b1ca294310b0bc/moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h#L67) in `setup_assistant/tools/moveit_config_data.h`. Their values are then loaded from the configuration file in the support package when running the setup assistant and added to the correct group in the `kinematics.yaml` file.

Does this mean we have to start thinking about releasing the plugin?

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 1 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: Well, this cache has no way of being invalidated by user-input so basically going back and changing things after running export once would not lead to updates in the exported data. I and probably also previous contributors thought this would only set static content (ie based on the urdf file not on user input) but looking at it this has not been true anyway.

I found an inbetween solution that I'll commit in a second. Re-exporting is still quite buggy visually but at least should not loose data.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch (hunk 3 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch]: Do you have a working setup were you could quickly test this? I have never used the benchmarking so far

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch (hunk 2 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch]: you're right, removed it

**Comment by gavanderhoorn** [comment]
> Does this mean we have to start thinking about releasing the plugin?

With the current implementation, the extra file is completely optional.

But from my point-of-view, I would say: yes, please release the plugin.

Don't let @davetcoleman know though: before you know it, he'll want to integrate it into `ros-planning/moveit` ;)

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch (hunk 3 above)
Comment (v4hn) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch]: Me neither.
I noticed `planning_context.launch` is included in this file too, so the parameters (in the `move_group` namespace) are set.
I'll forward the question to @henningkayser .

Comment (simonschmeisser) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch]: I tested the benchmark and loading the parameters here is indeed superfluous, removed it

**Comment by simonschmeisser** [comment]
@v4hn I changed this to master in order to follow the normal workflow, could you please do a review or tag someone who could/should ;) @davetcoleman is probably too busy to even read this tag

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 4]
```diff
@@ -1360,6 +1360,51 @@ bool MoveItConfigData::inputKinematicsYAML(const std::string& file_path)
...
+    ROS_ERROR("<group ns=\"$(arg robot_description)_kinematics\"> not found");
+    return false;
+  }
+
+  // iterate over all <rosparam namespace="group" file="..."/> elements
+  // and if 'group' matches an existing group, copy the filename
+  for (TiXmlElement* kinematics_parameter_file = kinematics_group->FirstChildElement("rosparam");
+       kinematics_parameter_file; kinematics_parameter_file = kinematics_parameter_file->NextSiblingElement("rosparam"))
+  {
+    const char* ns = kinematics_parameter_file->Attribute("ns");
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1398]: I don't get the role of the `ns` attribute here, which doesn't exist in the template file:
https://github.com/ros-planning/moveit/blob/4d3062164bf03ebb74c57d5eba0e3e50438481ba/moveit_setup_assistant/templates/moveit_config_pkg_template/launch/planning_context.launch#L19-L22
Shouldn't you check for the existence of the `file` attribute instead?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 5]
```diff
@@ -1360,6 +1360,51 @@ bool MoveItConfigData::inputKinematicsYAML(const std::string& file_path)
...
+  if (!launch_document.LoadFile())
+  {
+    ROS_ERROR_STREAM("Failed parsing " << file_path);
+    return false;
+  }
+
+  // find the kinematics section
+  TiXmlHandle doc(&launch_document);
+  TiXmlElement* kinematics_group = doc.FirstChild("launch").FirstChild("group").ToElement();
+  if (!kinematics_group)
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1378]: The check and warning here are somewhat superfluous given the check below.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 6]
```diff
@@ -1550,6 +1595,86 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
...
 
+// ******************************************************************************************
+// Extract the package/stack name from an absolute file path
+// Input:  path
+// Output: package name and relative path
+// ******************************************************************************************
+bool MoveItConfigData::extractPackageNameFromPath(const std::string& path, std::string& package_name,
+                                                  std::string& relative_filepath) const
+{
+  // Get the path to urdf, save filename
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1606]: Adapt the comment?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 7]
```diff
@@ -1550,6 +1595,86 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
...
+  fs::path sub_path;       // holds the directory less one folder
+  fs::path relative_path;  // holds the path after the sub_path
+
+  // Paths for testing if files exist
+  fs::path package_path;
+
+  std::vector<std::string> path_parts;  // holds each folder name in vector
+
+  // Copy path into vector of parts
+  for (fs::path::iterator it = file_directory.begin(); it != file_directory.end(); ++it)
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1620]: Use range-based for?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 8]
```diff
@@ -1550,6 +1595,86 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
...
+    // Reset the sub_path
+    sub_path.clear();
+
+    // Create a subpath based on the outer loops length
+    for (int segment_count = 0; segment_count < segment_length; ++segment_count)
+    {
+      sub_path /= path_parts[segment_count];
+
+      // decide if we should remember this directory name because it is topmost, in case it is the package/stack name
+      if (segment_count == segment_length - 1)
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1637]: Instead of checking this condition in the loop, you can move the assignment out from the loop:
`package_name = path_parts[segment_length - 1]`

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 9]
```diff
@@ -1550,6 +1595,86 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
...
+      }
+    }
+
+    // check if this directory has a package.xml
+    package_path = sub_path;
+    package_path /= "package.xml";
+    ROS_DEBUG_STREAM("Checking for " << package_path.make_preferred().string());
+
+    // Check if the files exist
+    if (fs::is_regular_file(package_path) || fs::is_regular_file(sub_path / "manifest.xml"))
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1649]: Could you make checking for package.xml and manifest.xml symmetric here?
The variable `package_path` isn't used other than (temporarily) holding `sub_path / "package.xml"`.
Please remove it.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 10]
```diff
@@ -1550,6 +1595,86 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
   return true;
 }
 
+// ******************************************************************************************
+// Extract the package/stack name from an absolute file path
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1599]: This whole function is rather complex. But I just have seen that you moved it from `StartScreenWidget`.
May I ask for some changes anyway?

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 11]
```diff
@@ -1118,6 +1121,18 @@ void ConfigurationFilesWidget::loadTemplateStrings()
     addTemplateString("[ROS_CONTROLLERS]", controllers.str());
   }
 
+  std::string kinematics_parameters_files_block;
```
Comment (rhaschke) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1124]: Can you add a comment to this section as for the others?

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch [hunk 12]
```diff
@@ -10,7 +10,6 @@
 
   <node name="$(anon rviz)" launch-prefix="$(arg launch_prefix)" pkg="rviz" type="rviz" respawn="false"
 	args="$(arg command_args)" output="screen">
-    <rosparam command="load" file="$(find [GENERATED_PACKAGE_NAME])/config/kinematics.yaml"/>
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: I guess, removing this line makes rviz not find the kinematics parameters anymore.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch [hunk 13]
```diff
@@ -15,7 +15,6 @@
 
   <!-- Start Benchmark Executable -->
   <node name="$(anon moveit_benchmark)" pkg="moveit_ros_benchmarks" type="moveit_run_benchmark" args="$(arg cfg) --benchmark-planners" respawn="false" output="screen">
-    <rosparam command="load" file="$(find [GENERATED_PACKAGE_NAME])/config/kinematics.yaml"/>
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch:18]: Same here. Did you test that these apps still find the kinematics parameters?

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/planning_context.launch [hunk 14]
```diff
@@ -19,6 +19,7 @@
   <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
   <group ns="$(arg robot_description)_kinematics">
     <rosparam command="load" file="$(find [GENERATED_PACKAGE_NAME])/config/kinematics.yaml"/>
+    [KINEMATICS_PARAMETERS_FILE_NAMES_BLOCK]
```
Comment (rhaschke) [inline moveit_setup_assistant/templates/moveit_config_pkg_template/launch/planning_context.launch:22]: See above
```suggestion
[KINEMATICS_PARAMETERS_FILE_NAMES_BLOCK]
```

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 15]
```diff
@@ -1118,6 +1121,18 @@ void ConfigurationFilesWidget::loadTemplateStrings()
...
+  for (const auto& groups : config_data_->group_meta_data_)
+  {
+    if (groups.second.kinematics_parameters_file_.empty())
+      continue;
+
+    std::string line = "<rosparam command=\"load\" ns=\"" + groups.first + "\" file=\"" +
```
Comment (rhaschke) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1130]: As far as I understand, it is possible to add multiple files here. However, only the first line is correctly indented (in the `planning_context.launch` template).
```suggestion
    std::string line = "    <rosparam command=\"load\" ns=\"" + groups.first + "\" file=\"" +
```

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch (hunk 12 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: planning_context is included which loads the parameters. Suggested by @v4hn , tested and still works fine without that line

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch (hunk 13 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/run_benchmark_ompl.launch:18]: Yes, even though the tutorial was somewhat hard to follow ... which I hope to address at some point

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 15 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1130]: Yes, one per planning group basically with the namespace (ns) being the group name

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 4 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp:1398]: the namespace is written in the block template "KINEMATICS_PARAMETERS_FILE_NAMES_BLOCK" (which is a bit ugly because there is no proper templating library with for loops like grantlee ...) and contains the name of the JointModelGroup

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch (hunk 12 above)
Comment (rhaschke) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: I don't see `planning_context.launch` being included here. It is, indeed, for `run_benchmark_ompl.launch`.

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 15 above)
Comment (rhaschke) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1130]: Does this new scheme (using the `ns` attribute) still work with old configs, which just declared all kinematics parameters in one single file?

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 10 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp:1599]: Sure, I just didn't want to put too much into one PR

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp (hunk 9 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp:1649]: I could also drop the manifest.xml ... that was in a world before my time ( > 5 years)

Comment (rhaschke) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp:1649]: Sure. I only noticed that you moved this code block after writing this comment.

Comment (rhaschke) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp:1649]: In any case, I suggest removing the variable package_path.

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 15 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1130]: yes, of course. You can simply leave the field for the additional kinematics parameter file on the group setting page empty and everything stays the same. But you could also choose the moveit_opw_kinematics plugin (or possibly in future UR kinematics) and an additional file with parameters distributed by ROS-I (or elsewhere). Then you would have a "old config" that just loads the correct plugin and new parameter files (from a description package) that add in the parameters.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch (hunk 12 above)
Comment (simonschmeisser) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: Agreed, but that file does not load other parameters either ... so my understanding was that it is not meant to be used standalone but rather as an include by other launchfiles?

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp (hunk 15 above)
Comment (rhaschke) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1130]: Thanks for your patience. I think I got the logic now.

Comment (simonschmeisser) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1130]: You're welcome! I promise to write a tutorial once we have the plugin released as well.

File: moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch (hunk 12 above)
Comment (rhaschke) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: I think you can use `moveit_rviz.launch` "standalone" to just load rviz, e.g. in a separate terminal. However, you are right that this usually assumes that `planning_context.launch` was launched somewhere else.
What I don't really understand right now, is why we used separate kinematics parameters for rviz in the past, but remove them now. Maybe @v4hn can help?

Comment (v4hn) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: For reasons we addressed over two years ago:

https://github.com/ros-planning/moveit/pull/701

Comment (rhaschke) [reply moveit_setup_assistant/templates/moveit_config_pkg_template/launch/moveit_rviz.launch:13]: Thanks for this link.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 16]
```diff
@@ -1603,65 +1598,25 @@ bool MoveItConfigData::setPackagePath(const std::string& pkg_path)
...
+  // truncate path step by step and check if it contains a package.xml
+  while (!sub_path.empty())
   {
-    // Reset the sub_path
-    sub_path.clear();
-
-    // Create a subpath based on the outer loops length
-    for (int segment_count = 0; segment_count < segment_length; ++segment_count)
+    ROS_DEBUG_STREAM("checking in " << sub_path.make_preferred().string());
+    if (fs::is_regular_file(sub_path / "package.xml") || fs::is_regular_file(sub_path / "manifest.xml"))
```
Comment (rhaschke) [inline moveit_setup_assistant/src/tools/moveit_config_data.cpp:1610]: Didn't you want to drop support for `manifest.xml`? I would strongly support this.

**Comment by v4hn** [comment]
Thanks for your work @simonschmeisser !
I'm glad we have you providing such nice additions and seeing them through. :smirk_cat:

**Comment by simonschmeisser** [comment]
I'd love to see this backported to melodic as well, what would I need to do for that?

**Comment by v4hn** [comment]
File a backport PR or wait for someone (we'll talk about it at the meeting later)
to look through all new patches in master and backport everything as needed). :-)

**Comment by davetcoleman** [comment]
> @davetcoleman is probably too busy to even read this tag

haha! i love these remarks. i catch up on my github emails every World MoveIt Day...
